rust/ip-packet/src/lib.rs

```diff
@@ -303,12 +303,34 @@ impl<'a> IpPacket<'a> {
         self.next_header() == IpNextHeaderProtocols::Tcp
     }
 
-    pub fn as_udp(&self) -> Option<UdpPacket> {
+    pub fn try_as_udp(&self) -> Option<UdpPacket> {
         self.is_udp()
             .then(|| UdpPacket::new(self.payload()))
             .flatten()
     }
 
+    pub fn as_udp_debug_checked(&self) -> Option<UdpPacket> {
+        debug_assert!(self.is_udp(), "Packet is not a UDP packet");
```

Comment:
At some point, to create the response to a dns we need the IP header, so we could copy the `UdpPacket` to an owned buffer so that we don't need to reparse, though pnet api requires cloning the underlying buffer and re-parse it so we still need this funciton.

Furthermore, then the function that construct the response would need to take both the `IpPacket` and the `UdpPacket` and the caller needs to be sure those matches, so it's easier to just re-parse.

Reply:
@thomaseizinger I'm not sure why we need to `debug_assert` here though, I think we should change the `unwrap` to `expect` on the caller to have this message.